File: ekosis/standard_endpoints/statistics.py
from pydantic import BaseModel as PydanticBaseModel

from ..requests.endpoint import endpoint
from ..state_keepers.statistics_keeper import StatisticsKeeper
from ..data_transfer_objects import StatsRequestDto, StatsResponseDto

# --------------------------------------------------------------------------------
@endpoint("eco.statistics.get", StatsRequestDto)
async def eco_statistics_get(dto: StatsRequestDto, **kwargs) -> PydanticBaseModel:
    stats_keeper = StatisticsKeeper()

    if dto.type == "current":
        stats = await stats_keeper.get_current_statistics()
    elif dto.type == "full":
        stats = await stats_keeper.get_full_gathered_statistics()
    else:
        stats = await stats_keeper.get_last_gathered_statistics()

    return StatsResponseDto(statistics=stats)

File: ekocpp/src/standard_endpoints/statistics.cpp
#include "statistics.hpp"

#include "../state_keepers/statistics_keeper.hpp"

namespace {

RequestDTO eco_statistics_get(SpanKey, RequestDTO& dto) {
    auto&             keeper = StatisticsKeeper::instance();
    const std::string type   = dto.data.value("type", std::string{"last"});

    if (type == "current") {
        return RequestDTO{{{"statistics", keeper.get_current_statistics()}}};
    }
    if (type == "full") {
        return RequestDTO{{{"statistics", keeper.get_full_gathered_statistics()}}};
    }
    return RequestDTO{{{"statistics", keeper.get_last_gathered_statistics()}}};
}

} // namespace

void register_statistics_endpoints(RequestRouter& router) {
    router.register_endpoint("eco.statistics.get", eco_statistics_get);
}
